### ISSUE TITLE
Implement a System for Representing Builtin Objects

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#4375]
This task is automatically imported from the old Task Issue Board and it was originally created by James Dunkerley.
Original issue is [here](https://www.pivotaltracker.com/story/show/181499077).
<hr />

Pivotal place holder for https://github.com/enso-org/enso/issues/1239
To flesh out with Marcin
(Not Basic Libraries but easier for JD to track!)

## Comments:
**Hubert Plociniczak** reports a new **STANDUP** for the last Friday (2022-05-06):

**Progress:** Was doing some background work for #181499077, had a chat with Marcin on next steps. Still some time was spent on addressing review for #182014385 (sbt for building libraries). It should be finished by 2022-05-23.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181499077). Address what appears to be a regression in builtins and continue on #181499077 (Enso Bot - May 9, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-05-09):

**Progress:** I've been debugging a regression in master which turned out to be related to delayed function body nodes' creation. Then started working on annotation processor for #181499077 for simple functions like Array.length. It should be finished by 2022-05-23.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181499077). Continue the work on the annotation processor and bring incremental improvements to our dsl. (Enso Bot - May 10, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-05-10):

**Progress:** Finished a first version of yet another annotation processor that generates BuiltinMethod nodes from method signatures. The latter gets processed by existing processors that generate the final classes. It should be finished by 2022-05-23.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181499077). Gathering feedback on the PR and continuing working on more examples. (Enso Bot - May 11, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-05-11):

**Progress:** Addressed feedback on the new processor. Added handling of simple constructors (`NewNode` class is now gone). In the meantime Radek discovered some issues with Frgaal and sbt. Apparently the latter didn't respect "provided" classifier and packaged it with other jars. Started looking into sbt build. It should be finished by 2022-05-23.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181499077). Address sbt build issues and continue on #181499077. (Enso Bot - May 12, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-05-12):

**Progress:** Working on the next PR that removes more boilerplate. Fixes issue in sbt with Frgaal being packaged as a dependency. It should be finished by 2022-05-23.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181499077). Continue work on dsl that removes more boilerplate code (Enso Bot - May 13, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for today (2022-05-13):

**Progress:** Finished another round of DSL improvements that generate more boilerplate code for builtins. Primarily focused on Array for now. PR is ready and awaiting review. It should be finished by 2022-05-23.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181499077). Address PR review comments and look for other cases when we can generate the code. (Enso Bot - May 13, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for today (2022-05-16):

**Progress:** Investigated potential next constructs to support. Started drafting support for Managed_Resource methods. Eventually realized that Managed_Resource has to undergo similar redesign in stdlib as Ref and then deal with functional dispatch boilerplate c&p. So I will park that work for now and address the latter first. It should be finished by 2022-05-23.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181499077). Address functional dispatch boilerplate  (e.g. include Array, Ref, Managed_Resource..) (Enso Bot - May 16, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-05-17):

**Progress:** Investigated builtin patterns and redesigned dsl to better represent the definition of builtins in more complicated cases It should be finished by 2022-05-23.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181499077). Continue work on builtins dsl (Enso Bot - May 18, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-05-18):

**Progress:** Fixed issues in separate compilation in the new dsl, continued adding new constructs. It should be finished by 2022-05-23.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181499077). Continue work on builtins dsl (Enso Bot - May 19, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-05-19):

**Progress:** Worked on builtins DSL for more complex examples (sort and copy for Array), will consult the draft design with Marcin before implementing it. After that moved to re-implementing EnsoFile with the builtins system. It should be finished by 2022-05-23.

**Next Day:** Next day I will be working on the [##181499077 task](https://www.pivotaltracker.com/story/show/181499077). Continue fixing any issues encountered with EnsoFile. (Enso Bot - May 20, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-05-19):

**Progress:** Continued on figuring out the missing bits of builtins dsl for redesign of File. The working solution with hardcoded method nodes works, now just need to work on the missing bits in the annotation processor. It should be finished by 2022-05-23.

**Next Day:** Next day I will be working on the [##181499077 task](https://www.pivotaltracker.com/story/show/181499077). Generate hardcoded methods necessary for File (Enso Bot - May 20, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for today (2022-05-20):

**Progress:** Continued on figuring out the missing bits of builtins dsl for redesign of File. The working solution with hardcoded method nodes works, now just need to work on the missing bits in the annotation processor. It should be finished by 2022-05-23.

**Next Day:** Next day I will be working on the [##181499077 task](https://www.pivotaltracker.com/story/show/181499077). Generate hardcoded methods necessary for File (Enso Bot - May 20, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-05-23):

**Progress:** WIP PR is up, shows what builtins dsl needs to model. Continued working on generating code. It should be finished by 2022-05-23.

**Next Day:** Next day I will be working on the [##181499077 task](https://www.pivotaltracker.com/story/show/181499077). Finish off generating code needed for File (Enso Bot - May 24, 2022)
<hr />
**Hubert Plociniczak** reports a new **🔴 DELAY** for yesterday (2022-05-24):

**Summary:** There is **4 days** delay in implementation of the [Implement a System for Representing Builtin Objects (#181499077) task](https://www.pivotaltracker.com/story/show/181499077).
It **will cause 0 days delay** for the delivery of this weekly plan.

**Delay Cause:** Initial plan was setup with a still unclear scope of the task. Work is going well (2 related PRs already merged) but it is just a lot of cases to cover.

**Possible solutions:** Divide bigger epics into smaller chunks. (Enso Bot - May 25, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-05-24):

**Progress:** Continued working on generating code. Investigated and fixed problems with (lack of) formatting for std-bits projects. Investigated issues with compilation of processors (it is a bit more involving, will need to schedule it separately). It should be finished by 2022-05-27.

**Next Day:** Next day I will be working on the [##181499077 task](https://www.pivotaltracker.com/story/show/181499077). Finish off generating code needed for File (Enso Bot - May 25, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-05-25):

**Progress:** A couple of 1:1s re next planning. Continued working on generating code for builtins (needed to do some refactoring first). It should be finished by 2022-05-27.

**Next Day:** Next day I will be working on the [##181499077 task](https://www.pivotaltracker.com/story/show/181499077). Finish off generating code needed for File (Enso Bot - May 26, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for today (2022-05-26):

**Progress:** All File builtins are now auto-generated from method signatures, including specializations. Added a bunch of cases for ManagedResource as well. It should be finished by 2022-05-27.

**Next Day:** Next day I will be working on the [##181499077 task](https://www.pivotaltracker.com/story/show/181499077). Cleanup documentation/code style in the pending PR, gather and address PR feedback (Enso Bot - May 26, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for the provided date (2022-05-27):

**Progress:** Cleaning up builtins PR. ready for review. Might need to delay it a bit to deal with any PR feedback. It should be finished by 2022-05-27.

**Next Day:** Next day I will be working on the [##181499077 task](https://www.pivotaltracker.com/story/show/181499077). Address PR feedback. (Enso Bot - May 30, 2022)
<hr />
**Hubert Plociniczak** reports a new **🔴 DELAY** for yesterday (2022-06-02):

**Summary:** There is **7 days** delay in implementation of the [Implement a System for Representing Builtin Objects (#181499077) task](https://www.pivotaltracker.com/story/show/181499077).
It **will cause 0 days delay** for the delivery of this weekly plan.

**Delay Cause:** Left the current state ready for review and picked up new work in the meantime, since it wasn't super-pressing. (Enso Bot - Jun 3, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-06-02):

**Progress:** Finally addressing PR comments related to builtins DSL (delayed by other work). Also, got feedback regarding short-circuiting boolean ops and simplified the whole solution. It should be finished by 2022-06-03.

**Next Day:** Next day I will be working on the [##181499077 task](https://www.pivotaltracker.com/story/show/181499077). Add more documentation, as was requested in comments. Pick up ticket next in backlog. (Enso Bot - Jun 3, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for the provided date (2022-06-03):

**Progress:** Continued cleaning up builtins DSL PR. Attempted to get rid of Prim_Warning as an exercise similar to File cleanup but that didn't go as planned. Ended up debugging with no success - will create a separate ticket as a follow up work. It should be finished by 2022-06-03.

**Next Day:** Next day I will be working on the [##181499077 task](https://www.pivotaltracker.com/story/show/181499077). Finish documentation. Pick up next ticket in backlog. (Enso Bot - Jun 6, 2022)
<hr />
**Hubert Plociniczak** reports a new **🔴 DELAY** for yesterday (2022-06-06):

**Summary:** There is **5 days** delay in implementation of the [Implement a System for Representing Builtin Objects (#181499077) task](https://www.pivotaltracker.com/story/show/181499077).
It **will cause 0 days delay** for the delivery of this weekly plan.

**Delay Cause:** Juggling a few issues at the same time, encountered some unexpected problems while applying dsl to more examples. (Enso Bot - Jun 7, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-06-06):

**Progress:** Decided to give Prim_Warning cleanup one last go and found the problem. Then continued the documentation effort. Looked at Scala compilation issues (related to #182359029). Had another look at ongoing language spec work. Various meetings. It should be finished by 2022-06-08.

**Next Day:** Next day I will be working on the [##181499077 task](https://www.pivotaltracker.com/story/show/181499077). Finish documentation. Pick up next ticket in backlog. (Enso Bot - Jun 7, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-06-07):

**Progress:** Got builtins DSL into shipable state. Added a ton of documentation. Moved on to other tasks while waiting on another round of review. It should be finished by 2022-06-08.

**Next Day:** Next day I will be working on the [##182209126 task](https://www.pivotaltracker.com/story/show/182209126). Addressing any review comments and looking into fixing instrumentation+incremental builds work started by Jaroslav (Enso Bot - Jun 8, 2022)
<hr />
